magefiles/testspecs/ginkgosspec.go

```diff
@@ -0,0 +1,205 @@
+var ginkgoGenerateSpecCmd = sh.OutCmd("ginkgo", "generate")
+
+// New returns a Ginkgo Spec Translator
+func NewGinkgoSpecTranslator() *GinkgosSpecTranslator {
+
+	return &GinkgosSpecTranslator{}
+}
+
+// FromFile generates a TestOutline from a Ginkgo test File
+func (gst *GinkgosSpecTranslator) FromFile(file string) (TestOutline, error) {
+
+	var nodes TestOutline
+	output, err := ginkgoOutlineJsonCmd(file)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal([]byte(output), &nodes)
+	if err != nil {
+		klog.Error("Failed ot unmarshal spec outline")
+		return nil, err
+	}
+	markInnerParentContainer(nodes)
+	nodes = excludeSetupTeardownNodes(nodes)
+	frameworkDescribeNode, err := ExtractFrameworkDescribeNode(file)
+	if err != nil {
+		klog.Error("Failed to extract the framework describe node from the AST")
+		return nil, err
+	}
+	if reflect.ValueOf(frameworkDescribeNode).IsZero() {
+		// we assume its a normal Ginkgo Spec file so return it
+		return nodes, nil
+	}
+	frameworkDescribeNode.Nodes = nodes
+	return TestOutline{frameworkDescribeNode}, nil
+}
+
+// ToFile generates a Ginkgo test file from a TestOutline
+func (gst *GinkgosSpecTranslator) ToFile(destination string, outline TestOutline) error {
+	testFilePath, err := createTestPath(destination)
+	if err != nil {
+		return err
+	}
+	dataFile, err := writeTemplateDataFile(testFilePath, outline)
+	if err != nil {
+		return err
+	}
+	return generateGinkgoSpec(testFilePath, dataFile)
+
+}
+
+// markInnerParentContainer marks whether
+// node is a parent container that comes after
+// the first root node which is the framework
+// describe decorator function
+func markInnerParentContainer(nodes TestOutline) {
+
+	for i := range nodes {
+		nodes[i].InnerParentContainer = true
+	}
+}
+
+// excludeSetupTeardownNodes removes those nodes from the ginkgo
+// outline output since they don't included anything useful anyways
+func excludeSetupTeardownNodes(nodes TestOutline) TestOutline {
+	excludes := []string{"JustBeforeEach", "BeforeEach", "BeforeAll", "JustAfterEach", "AfterAll", "AfterEach"}
+	for i := 0; i < len(nodes); i++ {
+		for _, ex := range excludes {
+			if ex == nodes[i].Name {
+				nodes = append(nodes[:i], nodes[i+1:]...)
+				nodes = excludeSetupTeardownNodes(nodes)
+				break
+			}
+
+		}
+
+		if len(nodes[i].Nodes) != 0 {
+			nodes[i].Nodes = excludeSetupTeardownNodes(nodes[i].Nodes)
+		}
+	}
+
+	return nodes
+
+}
+
+// generateGinkgoSpec will call the ginkgo generate command
+// to generate the ginkgo data json file we created and
+// the template located in out templates directory
+func generateGinkgoSpec(destination string, dataFile string) error {
+
+	var err error
+
+	// Note I change into the directory and rename things because ginkgo
+	// by default generates the test file name as <package>_test.go.
+	// Since that is not a semantic we follow I perform this action.
+	ginkgoFileName := fmt.Sprintf("%s_test.go", strings.Split(filepath.Dir(destination), "/")[1])
```

Comment:
If the destination is a full path, this code will not work as expected. I suggest something like:

```suggestion
dirs := strings.Split(filepath.Dir(destination), "/")
dir := dirs[len(dirs)-1]
ginkgoFileName := fmt.Sprintf("%s_test.go", dir)
```